src/main/java/net/minecraftforge/common/capabilities/INetworkCapability.java

```diff
@@ -0,0 +1,65 @@
+package net.minecraftforge.common.capabilities;
+
+import io.netty.buffer.Unpooled;
+import net.minecraft.network.FriendlyByteBuf;
+
+/**
+ * Provides the ability for {@link ICapabilityProvider}s to synchronize
+ * capabilities over the network.
+ */
+public interface INetworkCapability {
+
+    /**
+     * Write capabilities to the specified {@link FriendlyByteBuf}.
+     * 
+     * @param out      - the {@link FriendlyByteBuf} to write to
+     * @param writeAll - {@code true} if all data should be written, {@code false}
+     *                 if only dirty data should be written
+     */
+    void writeCapabilities(FriendlyByteBuf out, boolean writeAll);
+
+    /**
+     * Read capabilities from the specified {@link FriendlyByteBuf}.
+     * 
+     * @param in - {@link FriendlyByteBuf} to read from
+     */
+    void readCapabilities(FriendlyByteBuf in);
+
+    /**
+     * Whether capability data is 'dirty' and needs to be sent to the client.
+     */
+    boolean requiresSync();
+
+    /**
+     * A version of {@link #writeCapabilities(FriendlyByteBuf, boolean)} which
+     * returns a {@code byte[]}. This is intended for use with {@link ByteArrayTag},
+     * generally {@link #writeCapabilities(FriendlyByteBuf, boolean)} should be used
+     * to avoid unnecessarily creating a {@link byte[]}.
+     * 
+     * @param writeAll - {@code true} if all data should be written, {@code false}
+     *                 if only dirty data should be written
+     * @return a {@code byte[]} containing the serialized capability data
+     */
+    default byte[] writeCapabilities(boolean writeAll)
+    {
+        var buf = new FriendlyByteBuf(Unpooled.buffer());
```

Comment:
You wouldn't really be able to impose a size limit here unless you knew how much data was in the rest of the packet, assuming the caller of this even wants to put the data in a packet as it's not tied specially to that purpose.

Reply:
True enough, it could be limited to the max packet size, but at that point it really wouldn't matter since the erroring would always appear at the same place later down the line anyways. It's ore just for throwing errors near the 'correct' call site.